File: DSA/Cpp/test/dsa/lib/math/ds/palindrome_test.cpp
#include <dsa/lib/math/ds/palindrome.hpp>

namespace dsa::lib::math::ds {
namespace {

std::string getLongestPalindrome(std::string &&s) {
    auto r1 = Palindrome().longestPalindrome(s);
    auto r2 = Palindrome().longestPalindrome1(s);
    assert(r1 == r2 && "Results from 2 same functions are different.");

    return r1;
}
TEST(LongestPalindromeTest, emptyString) {
    EXPECT_EQ("", getLongestPalindrome(""));
}
TEST(LongestPalindromeTest, oneChar) {
    EXPECT_EQ("a", getLongestPalindrome("a"));
}
TEST(LongestPalindromeTest, case1) {
    EXPECT_EQ("aa", getLongestPalindrome("aa"));
    EXPECT_EQ("bab", getLongestPalindrome("babd"));
    EXPECT_EQ("bb", getLongestPalindrome("cbbd"));
    EXPECT_EQ("aaaa", getLongestPalindrome("aaaa"));
}

}  // namespace
}  // namespace dsa::lib::math::ds

File: DSA/Go/dsa/lib/math/ds/palindrome_test.go
package ds

import (
	"log"
	"testing"
)

func TestLongestPalindrome_emptyString(t *testing.T) {
	checkLongestPalindrome(getLongestPalindrome(""), "", t)
}

func TestLongestPalindrome_oneChar(t *testing.T) {
	checkLongestPalindrome(getLongestPalindrome("a"), "a", t)
}

func TestLongestPalindrome_case1(t *testing.T) {
	checkLongestPalindrome("aa", getLongestPalindrome("aa"), t)
	checkLongestPalindrome("bab", getLongestPalindrome("babd"), t)
	checkLongestPalindrome("bb", getLongestPalindrome("cbbd"), t)
	checkLongestPalindrome("aaaa", getLongestPalindrome("aaaa"), t)
}

func checkLongestPalindrome(want string, got string, t *testing.T) {
	if got != want {
		t.Errorf("getLongestPalindrome = %s, want = %s", got, want)
	}
}

func getLongestPalindrome(s string) string {
	r1 := longestPalindrome(s)
	r2 := longestPalindrome1(s)
	if r1 != r2 {
		log.Panicln("Results from 2 same functions are different.")
	}

	return r1
}
